<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/IDSNlogo.png" width="300" align="center">
</a>


<h1> HTTP and Requests</h1>

Estimated time needed: **30** minutes

## Objectives

After completing this lab you will be able to:

*   Understand HTTP
*   Handle HTTP Requests


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>
            <a href="#Overview-of-HTTP">Overview of HTTP </a>
            <ul>
                <li><a href="#Uniform-Resource-Locator:URL">Uniform Resource Locator:URL</a></li>
                 <li><a href="#Request">Request</a></li>
                <li><a href="#Response">Response</a></li>
            </ul>
        </li>
        <li>
            <a href="#Requests-in-Python">Requests in Python  </a>
            <ul>
                <li><a href="#Get-Request-with-URL-Parameters">Get Request with URL Parameters</a></li>
                <li><a href="#Post-Requests">Post Requests </a></li>

</ul>

</div>

<hr>


## Overview of HTTP


When you, the **client**, access a web page your browser sends an **HTTP** request to the **web server** that hosts the page. If no specific file is mentioned in the URL, the server typically looks for a default file, such as "<code>index.html</code>". If your request is successful, the server will send the object to the client in an **HTTP response**. This includes information like the type of the **resource**, the length of the **resource**, and other information.

<p>
The figure below represents the process. The circle on the left represents the client, the circle on the right represents the Web server. The table under the Web server represents a list of resources stored in the web server. In  this case an <code>HTML</code> file, <code>png</code> image, and <code>txt</code> file .
</p>
<p>
The <b>HTTP</b> protocol allows you to send and receive information through the web including webpages, images, and other web resources. In this lab, we will provide an overview of the Requests library for interacting with the <code>HTTP</code> protocol. 
</p


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/reqest_basics.png" width="750" align="center">

</div>


## Uniform Resource Locator:URL


Uniform resource locator (URL) is the most popular way to find resources on the web.  We can break the URL into three parts.

<ul>
    <li><b>Scheme</b>:- This is this protocol, for this lab it will always be <code>http://</code>  </li>
    <li><b> Internet address or  Base URL </b>:- This will be used to find the location here are some examples: <code>www.ibm.com</code> and  <code> www.gitlab.com </code> </li>
    <li><b>Route</b>:- Location on the web server for example: <code>/images/IDSNlogo.png</code> </li>
</ul>


You may also hear the term Uniform Resource Identifier (URI), URL are actually a subset of URIs. Another popular term is endpoint, this is the URL of an operation provided by a Web server.


## Request


The process can be broken into the <b>Request</b> and <b>Response </b> process.  The request using the get method is partially illustrated below. In the start line we have the <code>GET</code> method, this is an <code>HTTP</code> method. Also the location of the resource  <code>/index.html</code> and the <code>HTTP</code> version. The Request header passes additional information with an <code>HTTP</code> request:


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/reqest_messege.png" width="400" align="center">
</div>


When an <code>HTTP</code> request is made, an <code>HTTP</code> method is sent, this tells the server what action to perform.  A list of several <code>HTTP</code> methods is shown below. We will go over more examples later.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/http_methods.png" width="400" align="center">
</div>


## Response


The figure below represents the response; the response start line contains the version number <code>HTTP/1.0</code>, a status code (200) meaning success, followed by a descriptive phrase (OK). The response header contains useful information. Finally, we have the response body containing the requested file, an <code> HTML </code> document.  It should be noted that some requests have headers.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/response_message.png" width="400" align="center">
</div>


Some status code examples are shown in the table below, the prefix indicates the class. These are shown in yellow, with actual status codes shown in  white. Check out the following <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">link </a> for more descriptions.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/status_code.png" width="300" align="center">
</div>


## Requests in Python


Requests is a Python Library that allows you to send <code>HTTP/1.1</code> requests easily. We can import the library as follows:


In [1]:
import requests

We will also use the following libraries:


In [2]:
import os 
from PIL import Image
from IPython.display import IFrame

You can make a <code>GET</code> request via the method <code>get</code> to [www.ibm.com](http://www.ibm.com/?utm_source=Exinfluencer&utm_content=000026UJ&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01&utm_medium=Exinfluencer&utm_term=10006555):


In [3]:

url='https://www.ibm.com/'
r=requests.get(url)


We have the response object <code>r</code>, this has information about the request, like the status of the request. We can view the status code using the attribute <code>status_code</code>.


In [4]:
r.status_code

200

You can view the request headers:


In [5]:
print(r.request.headers)

{'User-Agent': 'python-requests/2.32.3', 'Accept-Encoding': 'gzip, deflate, br, zstd', 'Accept': '*/*', 'Connection': 'keep-alive', 'Cookie': '_abck=2972F31242FA980DA73AF9DC0B0D4839~-1~YAAQ7L7CF1Zlsq+XAQAAG8MpzA48Y59Pqf/UtROndrrQfu/LTQtFeXE/8LYx0eSRMsCJFuAe5mBF/+hmmhcxb6B9LlA6ANx8VuujlpcLzIVOXz5tDUDCm2CnwRNrqXRQF2TG6bG9bwP64ojCsOIJZV5ffP2IjTZjt5QG9+QoG53WQ5ELpKAoRddpRJUTQYcicHs/5NK/PWYozAkrBRM9rj6WbDaHD5tvbpvl6sXjIyJw72SDzXOvUJOW+UhCiZ2kbb6sTGYUKqsfuxOYJLBtsmf5tKjlLSRogsD6LluMtUBRvFaxeHliiC4U2awQoNCaKPFg/zpDt4Q1DPOvkUYc2O4/uNGim5UNiIH7Q9lYYEgkQOm9yfG2gcAUquK4bn+U+w7VOHoDGhPqxQBtDYAUOlfGckc0dDwA3R9JDxMUNJyoBXWTRBdW8p2EbupLDw==~-1~-1~-1; bm_sz=1CF42CD6F1B8D5111718FFE53B7433C7~YAAQ7L7CF1dlsq+XAQAAG8MpzBx05ZR+xrlzBTOyaoJy0WqVTTB4osSkBifPhZoz/WhQwpVhx9BITQWtk4cbZYYaH/fMO+0yIuM38PAjz1Ce6S8V2INq7w5YssPdDaFlpU9NAPoploiXyCfWXvzg4Z1O6Ead2Tc1OKnM3wjZ+FpRaSVbt3umbU+GhMIg6UwJCzWBBQ9GLtdN3l3UC8/nSfxjkBK4bBJZxsUB5mN9iWnFbnWJLS+GCQIiTiSnFhal38yA6KZ5VlCeGYH0HHlirZWgKbeqQdc2NqiMNO9rtfbb+8yAH0zyFYFE/k9Ql

You can view the request body, in the following line, as there is no body for a get request we get a <code>None</code>:


In [6]:
print("request body:", r.request.body)

request body: None


You can view the <code>HTTP</code> response header using the attribute <code>headers</code>. This returns a python dictionary of <code>HTTP</code> response headers.


In [7]:
header=r.headers
print(r.headers)

import pandas as pd
pd.set_option('display.max_columns', None)
header_df = pd.DataFrame(pd.Series(header))
header_df

{'Content-Security-Policy': 'upgrade-insecure-requests', 'x-frame-options': 'SAMEORIGIN', 'Last-Modified': 'Wed, 02 Jul 2025 17:21:53 GMT', 'ETag': '"2af00-638f57f406758-gzip"', 'Accept-Ranges': 'bytes', 'Content-Type': 'text/html;charset=utf-8', 'X-Content-Type-Options': 'nosniff', 'Cache-Control': 'max-age=600', 'Expires': 'Wed, 02 Jul 2025 17:32:58 GMT', 'X-Akamai-Transformed': '0 - 0 -', 'Content-Encoding': 'gzip', 'Date': 'Wed, 02 Jul 2025 17:22:58 GMT', 'Content-Length': '33172', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding', 'Strict-Transport-Security': 'max-age=31536000'}


0
Content-Security-Policy        upgrade-insecure-requests
x-frame-options                               SAMEORIGIN
Last-Modified              Wed, 02 Jul 2025 17:21:53 GMT
ETag                          "2af00-638f57f406758-gzip"
Accept-Ranges                                      bytes
Content-Type                     text/html;charset=utf-8
X-Content-Type-Options                           nosniff
Cache-Control                                max-age=600
Expires                    Wed, 02 Jul 2025 17:32:58 GMT
X-Akamai-Transformed                             0 - 0 -
Content-Encoding                                    gzip
Date                       Wed, 02 Jul 2025 17:22:58 GMT
Content-Length                                     33172
Connection                                    keep-alive
Vary                                     Accept-Encoding
Strict-Transport-Security               max-age=31536000

We can obtain the date the request was sent using the key <code>Date</code>.

In [8]:
header['Date']

'Wed, 02 Jul 2025 17:22:58 GMT'

<code>Content-Type</code> indicates the type of data:


In [9]:
header['Content-Type']

'text/html;charset=utf-8'

You can also check the <code>encoding</code>:


In [10]:
 r.encoding

'utf-8'

As the <code>Content-Type</code> is <code>text/html</code> we can use the attribute <code>text</code> to display the <code>HTML</code> in the body. We can review the first 100 characters:


In [11]:
r.text[0:100]

'\n<!DOCTYPE HTML>\n<html lang="en">\n<head>\r\n    \r\n    \r\n    \r\n    \r\n    \r\n    \r\n    \r\n      \r\n    \r\n  '

You can load other types of data for non-text requests, like images. Consider the URL of the following image:


In [12]:
# Use single quotation marks for defining string
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/IDSNlogo.png'

We can make a get request:


In [13]:
r=requests.get(url)

We can look at the response header:


In [14]:
print(r.headers)

picture_df = pd.DataFrame(pd.Series(r.headers))
picture_df

{'Date': 'Wed, 02 Jul 2025 17:22:59 GMT', 'X-Clv-Request-Id': '35c5c37e-3a95-4127-9b25-9108516b8751', 'Server': 'Cleversafe', 'X-Clv-S3-Version': '2.5', 'Accept-Ranges': 'bytes', 'x-amz-request-id': '35c5c37e-3a95-4127-9b25-9108516b8751', 'ETag': '"8bb44578fff8fdcc3d2972be9ece0164"', 'Content-Type': 'image/png', 'Last-Modified': 'Wed, 16 Nov 2022 03:32:41 GMT', 'Content-Length': '78776'}


0
Date                     Wed, 02 Jul 2025 17:22:59 GMT
X-Clv-Request-Id  35c5c37e-3a95-4127-9b25-9108516b8751
Server                                      Cleversafe
X-Clv-S3-Version                                   2.5
Accept-Ranges                                    bytes
x-amz-request-id  35c5c37e-3a95-4127-9b25-9108516b8751
ETag                "8bb44578fff8fdcc3d2972be9ece0164"
Content-Type                                 image/png
Last-Modified            Wed, 16 Nov 2022 03:32:41 GMT
Content-Length                                   78776

We can see the <code>'Content-Type'</code>


In [15]:
r.headers['Content-Type']

'image/png'

An image is a response object that contains the image as a <a href="https://docs.python.org/3/glossary.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01#term-bytes-like-object">bytes-like object</a>. As a result, we must save it using a file object. First, we specify the <u>file path and
name</u>


In [16]:
path=os.path.join(os.getcwd(),'image.png')
help(os.path.join)

Help on function join in module posixpath:

join(a, *p)
    Join two or more pathname components, inserting '/' as needed.
    If any component is an absolute path, all previous path components
    will be discarded.  An empty last part will result in a path that
    ends with a separator.



We save the file, in order to access the body of the response we use the attribute <code>content</code> then save it using the <code>open</code> function and write <code>method</code>:


In [17]:
with open(path,'wb') as f:
    f.write(r.content)

We can view the image:


In [18]:
Image.open(path)
help(Image)

Help on module PIL.Image in PIL:

NAME
    PIL.Image

DESCRIPTION
    # The Python Imaging Library.
    # $Id$
    #
    # the Image class wrapper
    #
    # partial release history:
    # 1995-09-09 fl   Created
    # 1996-03-11 fl   PIL release 0.0 (proof of concept)
    # 1996-04-30 fl   PIL release 0.1b1
    # 1999-07-28 fl   PIL release 1.0 final
    # 2000-06-07 fl   PIL release 1.1
    # 2000-10-20 fl   PIL release 1.1.1
    # 2001-05-07 fl   PIL release 1.1.2
    # 2002-03-15 fl   PIL release 1.1.3
    # 2003-05-10 fl   PIL release 1.1.4
    # 2005-03-28 fl   PIL release 1.1.5
    # 2006-12-02 fl   PIL release 1.1.6
    # 2009-11-15 fl   PIL release 1.1.7
    #
    # Copyright (c) 1997-2009 by Secret Labs AB.  All rights reserved.
    # Copyright (c) 1995-2009 by Fredrik Lundh.
    #
    # See the README file for information on usage and redistribution.
    #

CLASSES
    builtins.Exception(builtins.BaseException)
        DecompressionBombError
    builtins.RuntimeWarning(buil

<h3>Question: Download a file </h3>


Consider the following URL.


<code>URL = <https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt</code>


Write the commands to download the txt file in the given link.


In [19]:
URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt'
r=requests.get(URL)

header_df = pd.DataFrame(pd.Series(r.headers))
header_df

r.text
with open(os.path.join(os.getcwd(),'mandatory_assignment.txt'),'wt') as writefile:
    writefile.write(r.text)

r.content

b'This is line 1 \nThis is line 2\nThis is line 3'

<details><summary>Click here for the solution</summary>

```python
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt'
path=os.path.join(os.getcwd(),'example1.txt')
r=requests.get(url)
with open(path,'wb') as f:
    f.write(r.content)

```

</details>


## Get Request with URL Parameters


You can use the <b>GET</b> method to modify the results of your query, for example retrieving data from an API. We send a <b>GET</b> request to the  server. Like before we have the <b>Base URL</b>, in the <b>Route</b> we append <code>/get</code>, this indicates we would like to preform a <code>GET</code> request. 


The Base URL is for <code>[http://httpbin.org/](http://httpbin.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01)</code> is a simple HTTP Request & Response Service. The <code>URL</code> in Python is given by:


In [20]:
url_get='http://httpbin.org/get'

A <a href="https://en.wikipedia.org/wiki/Query_string?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">query string</a> is a part of a uniform resource locator (URL), this sends other information to the web server. The start of the query is a <code>?</code>, followed by a series of parameter and value pairs, as shown in the table below. The first parameter name is <code>name</code> and the value is <code>Joseph</code>. The second parameter name is <code>ID</code> and the Value is <code>123</code>. Each pair, parameter, and value is separated by an equals sign, <code>=</code>.
The series of pairs is separated by the ampersand <code>&</code>.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/query_string.png" width="500" align="center">
</div>


To create a Query string, add a dictionary. The keys are the parameter names and the values are the value of the Query string.


In [21]:
payload={"name":"Joseph","ID":"123"}

Then passing the dictionary <code>payload</code> to the <code>params</code> parameter of the <code> get()</code> function:


In [22]:
r=requests.get(url_get,params=payload)

We can print out the <code>URL</code> and see the name and values.


In [23]:
r.url

'https://www.safebrowse.io/warn.html?url=http://httpbin.org/get?name=Joseph&ID=123&token=9fb6b02f'

There is no request body.


In [24]:
print("request body:", r.request.body)

request body: None


We can print out the status code.


In [25]:
print(r.status_code)

200


We can view the response as text:


In [26]:
print(r.text)

<!doctype html>
<html lang="en">

<head>
  <meta name="referrer" content="unsafe-url">
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>Potential Threat Detected</title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="msapplication-TileColor" content="#000000">
  <meta name="msapplication-config" content="./browserconfig.xml">
  <meta name="theme-color" content="#000000">

  <link rel="stylesheet" href="./css/theme-xdns-security.min.css" inline>

  <script type="text/javascript" src="js/jquery/jquery.js"></script>
  <script type="text/javascript" src="js/class/class.min.js"></script>
  <script type="text/javascript" src="js/jquery-encoder/jquery.jquery-encoder.min.js"></script>
  <script type="text/javascript" src="js/dom-purify/purify.min.js"></script>
  <script type="text/javascript" src="js/warn.js"></script>
</head>

<body onload="render()">
  <header class="header">
    <img src="./img/icon_enha

We can look at the <code>'Content-Type'</code>.


In [27]:
r.headers['Content-Type']

'text/html'

As the content <code>'Content-Type'</code> is in the <code>JSON</code> format we can use the method <code>json()</code>, it returns a Python <code>dict</code>:


In [28]:
r.json()

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The key <code>args</code> has the name and values:


In [29]:
r.json()['args']

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

## Post Requests


Like a <code>GET</code> request, a <code>POST</code> is used to send data to a server, but the <code>POST</code> request sends the data in a request body. In order to send the Post Request in Python, in the <code>URL</code> we change the route to <code>POST</code>:


In [ ]:
url_post='http://httpbin.org/post'

This endpoint will expect data as a file or as a form. A form is convenient way to configure an HTTP request to send data to a server.


To make a <code>POST</code> request we use the <code>post()</code> function, the variable <code>payload</code> is passed to the parameter <code> data </code>:


In [ ]:
r_post=requests.post(url_post,data=payload)

Comparing the URL from the response object of the <code>GET</code> and <code>POST</code> request we see the <code>POST</code> request has no name or value pairs.


In [ ]:
#print("POST request URL:",response.url )
#print("GET request URL:",r.url)
print("POST request URL:", r_post.url)  # Use r_post instead of response

We can compare the <code>POST</code> and <code>GET</code> request body, we see only the <code>POST</code> request has a body:


In [ ]:
print("POST request body:",r_post.request.body)
print("GET request body:",r.request.body)

We can view the form as well:


In [ ]:
r_post.json()['form']

There is a lot more you can do. Check out <a href="https://requests.readthedocs.io/en/master/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Requests </a> for more.


<hr>


## Authors

<p><a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01" target="_blank">Joseph Santarcangelo</a> <br>A Data Scientist at IBM, and holds a PhD in Electrical Engineering. His research focused on using Machine Learning, Signal Processing, and Computer Vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.</p>


### Other Contributors

<a href="https://www.linkedin.com/in/jiahui-mavis-zhou-a4537814a?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Mavis Zhou</a>


## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
<!--
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description           |
| ----------------- | ------- | ---------- | ---------------------------- |
| 2023-11-02 | 2.4 | Abhishek Gagneja | Updated instructions |
| 2023-06-07        | 2.3     |Akansha Yadav| Spell Check                 |
| 2021-12-20        | 2.1     | Malika     | Updated the links            |
| 2020-09-02        | 2.0     | Simran     | Template updates to the file |
|                   |         |            |                              |
|                   |         |            |                              |

--!>
